sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -216,10 +216,16 @@ class Analyzer(
      *  Group Count: N + 1 (N is the number of group expressions)
      *
      *  We need to get all of its subsets for the rule described above, the subset is
-     *  represented as the bit masks.
+     *  represented as sequence of expressions.
      */
-    def bitmasks(r: Rollup): Seq[Int] = {
-      Seq.tabulate(r.groupByExprs.length + 1)(idx => (1 << idx) - 1)
+    def rollupExprs(exprs: Seq[Expression]): Seq[Seq[Expression]] = {
+      val buffer = ArrayBuffer.empty[Seq[Expression]]
```

Comment:
Avoid using `ArrayBuffer` as insertions would lead to expansion of underlying array and copying of data to the new one. Since you know the size upfront, you could create an `Array` of required size.